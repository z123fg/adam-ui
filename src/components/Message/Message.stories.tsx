import React from "react";
import { ComponentStory, ComponentMeta } from '@storybook/react';
import Message, {
  MessageSize,
  MessageColor,
  MessageType
} from "./Message";

export default {
    title: "Message",
    component: Message
}as ComponentMeta<typeof Message>;

const Template: ComponentStory<typeof Message> = (args) => <Message {...args} />;

export const Type = Template.bind({});
Type.args = {
  msgType: MessageType.Urgent,
  msgContent: 'Message',
};

export const Color = Template.bind({});
Color.args = {
  msgColor:MessageColor.Red,
  msgContent: 'Message',
};

export const Size = Template.bind({});
Size.args = {
  msgSize: MessageSize.Large,
  msgContent: 'Message',
};

export const Click = Template.bind({});
Click.args = {
  msgClick: true,
  msgContent: 'Message',
};

export const CustomMessage = Template.bind({});
CustomMessage.args = {
  msgContent: 'Custom Message',
};