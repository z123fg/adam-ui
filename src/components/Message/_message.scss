.msg {
  vertical-align: middle;
  position: relative;
  // cursor: pointer;
  padding: 1.25rem 1.25rem 0.75rem 0.75rem;
  border: 1px solid black;
  border-radius: 0.25rem;
  width: fit-content;
  font-size: 16;
}

.msg-lg {
  font-size: x-large;
}

.msg-sm {
  font-size: medium;
}

.msg-default {
  background-color: white;
}

.msg-urgent {
  border-width: thick;
  font-weight: 800;
  // background-color: #ff5757;
  // color: #b30000;
  // border-color: #ff0606;
}

.msg-red {
  color: #c83349;
  background-color: #f8d7da;
  border-color: #e99b9a;
}

.msg-black {
  color: black;
}

.msg-blue {
  background-color: #78b9e7;
  color: #004c91;
  border-color: #007dc6;
}

.out-top {
  position: fixed;
  width: 100%;
  top: 0;
  left: 0;
}
.msg-top {
  position: absolute;
}

.out-bottom {
  position: fixed;
  width: 100%;
  bottom: 5;
  left: 5;
}
.msg-bottom {
  position: absolute;
  // bottom: 2%;
  // left: 2%;
}
