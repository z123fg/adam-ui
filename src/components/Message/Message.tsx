import { padding } from '@mui/system';
import React, { FC, useState } from 'react';
import genClassName from "../../util/genClassName";
import "./_message.scss"

//big small message
export enum MessageSize {
  Large = "lg",
  Small = "sm",
}

//text color change
export enum MessageColor {
  Black = "black",
  Red = "red",
  Blue = "blue",
}

//background and border color change
export enum MessageType {
  Urgent = "urgent",
  Default = "default",
}

interface IMessagesProps {
    className?: string;
    msgType?: MessageType;
    msgSize?: MessageSize;
    msgColor?: MessageColor;
    msgContent?: string;
    msgClick?: boolean; //false for a auto delete message(alert), true for click delete
}

export const Message: FC<IMessagesProps> = ({
    className = "", 
    msgType=MessageType.Default,
    msgSize=MessageSize.Small, 
    msgColor = MessageColor.Black,
    msgContent = "",
    msgClick=false,
        ...rest 
}) => {

    let classNameStr = genClassName(className,
        "msg",
        {
            [`msg-still`]:!!msgClick,
        [`msg-${msgType}`]: !!msgType,
        [`msg-${msgSize}`]: !!msgSize,
        [`msg-${msgColor}`]: !!msgColor,
    }  ).join(" ");


    const [show, setShow] = useState(msgClick);
    const [appear, setAppear] = useState(true);

    let autoMsg = (
        <div className={classNameStr} {...(rest as IMessagesProps)}
            data-testid='message-element'>
            <div className='content'>
                <p>{msgContent}</p>
            </div>
        </div >    
    )
        
    setTimeout(() => {
        setAppear(false)
    }, 2000);

    if (msgClick) {
        if (show) {
            return (
                <div className={classNameStr} {...(rest as IMessagesProps)}
                    data-testid='message-element'
                    onClick={(e) => {
                        e.preventDefault();
                        setShow(false);
                    }}>
                    <div style={{ textAlign: "right", margin: "-10px 0px -5px 0px", cursor:"pointer" }}>X</div>
                    <div className='content'>
                        <p>{msgContent}</p>
                    </div>
                </div >        
            );
        } else return null
    }

    if(appear) {
        return autoMsg;
    }

    return (
        null
    )

};

export default Message;