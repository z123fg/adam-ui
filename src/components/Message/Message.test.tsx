import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Message, { MessageSize, MessageColor, MessageType } from "./Message";


describe("test Message Component", () => {
    it("should snapshot", () =>{
        const wrapper = render(<Message>Default Message</Message>);
        const firstRender = wrapper.asFragment();
        expect(firstRender).toMatchSnapshot();
        expect(firstRender).toMatchInlineSnapshot(`
        <DocumentFragment>
            <div class='content'>
                <p>Default Message</p>
            </div>
        </DocumentFragment>
        `)
    })
  it("should render without issue", () => {
    const wrapper = render(<Message>Default Message</Message>);
    const messageElement = wrapper.queryByText("Default Message");
    expect(messageElement).toBeInTheDocument();
  });

//   it("should render correct button based on the prop", () => {
//     const mockOnClickHandler = jest.fn(); //it creates a mock up function
//     const buttonProps = {
//       messageColor: MessageColor.Black,
//       messageSize: MessageSize.Large,
//     };
//     const wrapper = render(
//       <Message
//         {...buttonProps}
//       >
//         Message color and size
//       </Message>
//     );
//     const buttonElement = wrapper.queryByText("Button type");
//     expect(buttonElement).toBeInTheDocument();
//     expect(buttonElement?.tagName).toBe("BUTTON");
//     expect(buttonElement).toHaveClass(
//       "msg",
//       "msg-urgent",
//       "msg-lg",
//       "msg-red"
//     );
//     expect(mockOnClickHandler).toBeCalledTimes(0);
//     buttonElement && fireEvent.click(buttonElement);
//     expect(mockOnClickHandler).toBeCalledTimes(1);
//   });

//   it("Urgent button should be rendered properly based on props", () => {
//     const wrapper = render(
//       <Message messageType={MessageType.Urgent}>Urgent Button</Message>
//     );
//     //const buttonElement = wrapper.queryByRole("Urgent",{name:"Urgent Button"});
//     const buttonElement = wrapper.queryByText("Urgent Button");
//     expect(buttonElement).toHaveClass("msg", "msg-urgent", "msg-sm");
//   });

//   it("Urgent button should have be disabled", () => {
//     const mockOnClickHandler = jest.fn();
//     const wrapper = render(
//       <Message
//         MessageType={MessageType.Urgent}
//       >
//         Urgent Button
//       </Message>
//     );

//     const buttonElement = wrapper.queryByText("Urgent Button");
//     expect(mockOnClickHandler).toBeCalledTimes(0);
//     buttonElement && fireEvent.click(buttonElement);
//     expect(mockOnClickHandler).toBeCalledTimes(0);
//     expect(buttonElement).toHaveClass("msg", "msg-Urgent", "msg-sm", "disabled");
//   });
});
