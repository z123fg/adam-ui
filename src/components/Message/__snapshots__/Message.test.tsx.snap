// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Message Component should snapshot 1`] = `
<DocumentFragment>
  <div
    class="msg  msg-default msg-sm msg-black"
    data-testid="message-element"
  >
    <div
      class="content"
    >
      <p />
    </div>
  </div>
</DocumentFragment>
`;
